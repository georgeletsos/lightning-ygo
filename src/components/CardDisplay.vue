<template>
  <div v-if="Object.keys(displayCard).length > 0" class="h-100">
    <div id="l-card-image">
      <font-awesome-icon
        v-if="displayCards.length > 1"
        class="l-prev animated infinite bounceLeft faster"
        :icon="faIcons.faCaretLeft"
        size="3x"
        @click="fetchPrevDisplayCard(displayCard)"
      />
      <font-awesome-icon
        v-if="displayCards.length > 1"
        class="l-next animated infinite bounceRight faster"
        :icon="faIcons.faCaretRight"
        size="3x"
        @click="fetchNextDisplayCard(displayCard)"
      />

      <transition :enter-active-class="enterActiveClass" mode="out-in">
        <img
          v-if="showArtImage"
          key="1"
          :src="displayCard.image.art"
          :alt="displayCard.name"
          class="img-fluid l-image-art"
          @click="
            showArtImage = false;
            enterActiveClass = 'animated zoomOut faster';
          "
        />
        <img
          v-else
          key="2"
          :src="displayCard.image.big"
          :alt="displayCard.name"
          class="img-fluid l-image-big"
          @click="
            showArtImage = true;
            enterActiveClass = 'animated zoomIn faster';
          "
        />
      </transition>
    </div>
    <div id="l-card-data">
      <div id="l-card-name-level-attr">
        <div class="l-card-info-box">
          <div class="l-card-info-inner-box">
            <div class="card-name">{{ displayCard.name }}</div>
            <hr />
            <div class="d-flex">
              <div
                v-if="isMonsterCard(displayCard)"
                class="l-monster-level"
                :class="{
                  'l-tuner-monster-level': isTunerMonsterCard(displayCard)
                }"
              >
                <img class="img-fluid" :src="monsterLevelIcon" alt="Level" />{{
                  displayCard.level
                }}
              </div>
              <div v-else-if="isSpellCard(displayCard)" class="l-st-card-type">
                <img class="img-fluid" :src="spellIcon" alt="Card Type" />{{
                  displayCard.cardType | capitalize
                }}
              </div>
              <div v-else-if="isTrapCard(displayCard)" class="l-st-card-type">
                <img class="img-fluid" :src="trapIcon" alt="Card Type" />{{
                  displayCard.cardType | capitalize
                }}
              </div>

              <div v-if="isMonsterCard(displayCard)" class="l-monster-attr">
                <img
                  class="img-fluid"
                  :src="getMonsterAttributeIcon(displayCard.attribute)"
                  alt="Attribute"
                />{{ displayCard.attribute.toUpperCase() }}
              </div>
              <div
                v-else-if="
                  (isSpellCard(displayCard) || isTrapCard(displayCard)) &&
                    !isStNormal(displayCard)
                "
                class="l-st-type"
              >
                <img
                  class="img-fluid"
                  :src="getStTypeIcon(displayCard.types[0])"
                  alt="Spell/Trap Type"
                />{{ displayCard.types[0] | capitalize }}
              </div>
            </div>
          </div>
        </div>
      </div>
      <div id="l-card-type-text-stats">
        <div class="l-card-info-box">
          <div class="l-card-info-inner-box">
            <div v-if="isMonsterCard(displayCard)" class="l-monster-types">
              [{{ capitalizedDisplayCardMonsterType }}/{{
                displayCard.types
                  .map(t => $options.filters.capitalize(t))
                  .join("/")
              }}]
            </div>
            <div
              ref="cardText"
              class="l-card-text"
              :class="{
                'l-full-height':
                  isSpellCard(displayCard) || isTrapCard(displayCard)
              }"
            >
              {{ displayCard.text }}
            </div>
            <div v-if="isMonsterCard(displayCard)" class="l-monster-stats">
              {{ `ATK/${displayCard.atk} DEF/${displayCard.def}` }}
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";
import { faCaretLeft, faCaretRight } from "@fortawesome/free-solid-svg-icons";
import { mapGetters } from "vuex";
import {
  FETCH_PREV_DISPLAY_CARD,
  FETCH_NEXT_DISPLAY_CARD
} from "@/store/actions.type";
import {
  isMonsterCard,
  isTunerMonsterCard,
  isSpellCard,
  isTrapCard,
  isStNormal
} from "@/common/utilities";
import {
  spellIcon,
  trapIcon,
  monsterLevelIcon,
  getMonsterAttributeIcon,
  getStTypeIcon
} from "@/common/assets";

export default {
  components: {
    FontAwesomeIcon
  },

  computed: {
    faIcons() {
      return {
        faCaretLeft,
        faCaretRight
      };
    },
    capitalizedDisplayCardMonsterType() {
      const displayCardMonsterType = this.displayCard.monsterType;

      for (const separator of [" ", "-"]) {
        if (displayCardMonsterType.indexOf(separator) > -1) {
          return displayCardMonsterType
            .split(separator)
            .map(word => this.$options.filters.capitalize(word))
            .join(separator);
        }
      }

      return this.$options.filters.capitalize(displayCardMonsterType);
    },
    spellIcon: () => spellIcon,
    trapIcon: () => trapIcon,
    monsterLevelIcon: () => monsterLevelIcon,
    ...mapGetters(["displayCard", "displayCards"])
  },

  watch: {
    // When the display card changes, reset the card text scroll
    displayCard() {
      if (this.$refs.cardText) {
        this.$refs.cardText.scrollTop = 0;
      }
    }
  },

  data() {
    return {
      showArtImage: false,
      enterActiveClass: ""
    };
  },

  filters: {
    capitalize: function(value) {
      if (!value) return "";
      value = value.toString();
      return value.charAt(0).toUpperCase() + value.slice(1);
    }
  },

  methods: {
    fetchPrevDisplayCard(displayCard) {
      this.$store.dispatch(FETCH_PREV_DISPLAY_CARD, displayCard);
    },
    fetchNextDisplayCard(displayCard) {
      this.$store.dispatch(FETCH_NEXT_DISPLAY_CARD, displayCard);
    },
    isMonsterCard,
    isTunerMonsterCard,
    isSpellCard,
    isTrapCard,
    isStNormal,
    getMonsterAttributeIcon,
    getStTypeIcon
  }
};
</script>

<style lang="scss" scoped>
$blue: #001569;
$green: #6ed96c;
$light-blue-100: #0d659b;
$light-blue-200: #007fb4;
$light-blue-300: #003f89;
$light-blue-400: #003773;
$light-blue-500: #0055b7;
$white: #ffffff;

#l-card-image {
  height: 40%;
  overflow: hidden;
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: center;
  user-select: none;

  img {
    max-height: 100%;
    display: block;
    margin: 0 auto;
    padding-bottom: 0.5rem;

    &.l-image-big {
      cursor: zoom-in;
    }

    &.l-image-art {
      cursor: zoom-out;
    }
  }

  .l-prev,
  .l-next {
    cursor: pointer;
    color: $white;
  }

  .l-prev {
    position: absolute;
    left: 5px;
  }

  .l-next {
    position: absolute;
    right: 5px;
  }
}

.zoomIn {
  animation-name: zoomIn;
}

.zoomOut {
  animation-name: zoomOut;
}

.bounceLeft {
  animation-name: bounceLeft;
}

.bounceRight {
  animation-name: bounceRight;
}

@keyframes zoomIn {
  0% {
    transform: scale(0.525) translateY(-26px);
  }
  100% {
    transform: scale(1);
  }
}

@keyframes zoomOut {
  0% {
    transform: scale(1.9) translateY(14px);
  }
  100% {
    transform: scale(1);
  }
}

@keyframes bounceLeft {
  0%,
  100% {
    transform: translateX(0);
  }

  50% {
    transform: translateX(-5px);
  }
}

@keyframes bounceRight {
  0%,
  100% {
    transform: translateX(0);
  }

  50% {
    transform: translateX(5px);
  }
}

#l-card-data {
  height: 60%;

  .l-card-info-box {
    height: 100%;
    background-color: $light-blue-300;
    border: 2px solid $light-blue-500;
    padding: 0.25rem;
    color: $white;

    .l-card-info-inner-box {
      height: 100%;
      background-color: $light-blue-400;
      padding: 0.25rem;
    }
  }
}

#l-card-name-level-attr {
  margin-bottom: 0.5rem;

  hr {
    margin: 0.25rem 3rem;
    border: 0;
    border-top: 1px solid $light-blue-100;
  }

  .l-monster-level,
  .l-monster-attr,
  .l-st-card-type,
  .l-st-type {
    display: flex;
    align-items: center;

    img {
      height: 24px;
      margin-right: 0.25rem;
    }
  }

  .l-monster-level,
  .l-st-card-type {
    margin-right: 1rem;
  }

  .l-tuner-monster-level {
    color: $green;
  }
}

#l-card-type-text-stats {
  // #l-card-name-level-attr has 77px height + 0.5rem margin-bottom
  height: calc(100% - (77px + 0.5rem));

  .l-monster-types {
    color: $green;
  }

  .l-card-text {
    // .l-monster-types has 24px height + .l-monster-stats has 40px height
    height: calc(100% - (24px + 40px));
    overflow-y: scroll;
    white-space: pre-line;

    &.l-full-height {
      height: 100%;
    }
  }

  .l-monster-stats {
    border-top: 2px solid $light-blue-200;
    background-color: $blue;
    padding: 0.25rem;
    margin: 0.25rem -0.25rem 0 -0.25rem;
    font-size: 1.25rem;
    text-align: right;
  }
}
</style>
