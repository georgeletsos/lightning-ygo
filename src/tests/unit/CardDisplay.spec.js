import { mount, createLocalVue } from "@vue/test-utils";
import flushPromises from "flush-promises";
import CardDisplay from "@/components/CardDisplay.vue";
import Vuex from "vuex";
import {
  FETCH_PREV_DISPLAY_CARD,
  FETCH_NEXT_DISPLAY_CARD
} from "@/store/actions.type";
import { mockCards } from "./mocks";

const localVue = createLocalVue();
localVue.use(Vuex);

localVue.directive("touch", () => {});

const actions = {
  [FETCH_PREV_DISPLAY_CARD]: () => {},
  [FETCH_NEXT_DISPLAY_CARD]: () => {}
};
const getters = {
  displayCard: () => mockCards[2],
  displayCards: () => mockCards
};
const store = new Vuex.Store({
  actions,
  getters
});

describe("CardDisplay.vue", () => {
  it("should render correctly", () => {
    const wrapper = mount(CardDisplay, { store, localVue });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should test if the |capitalizedDisplayCardMonsterType| computed property works correctly", () => {
    const wrapper = mount(CardDisplay, { store, localVue });
    expect(wrapper.vm.capitalizedDisplayCardMonsterType).toBe("Machine");
  });

  it("should test if the |existsInDisplayCards| method works correctly", () => {
    const wrapper = mount(CardDisplay, { store, localVue });
    expect(wrapper.vm.existsInDisplayCards(wrapper.vm.displayCard)).toBe(true);
    expect(
      wrapper.vm.existsInDisplayCards({ name: "Ancient Gear Golem" })
    ).toBe(false);
  });

  it("should switch between the card image/art image, on card image/art image click", async () => {
    const wrapper = mount(CardDisplay, { store, localVue });
    expect(wrapper.find(".l-image-big").exists()).toBe(true);

    await wrapper.find(".l-image-big").trigger("click");
    expect(wrapper.find(".l-image-big").exists()).toBe(false);
    expect(wrapper.find(".l-image-art").exists()).toBe(true);

    await wrapper.find(".l-image-art").trigger("click");
    expect(wrapper.find(".l-image-art").exists()).toBe(false);
    expect(wrapper.find(".l-image-big").exists()).toBe(true);
  });

  it("should call the |fetchPrevDisplayCard| method, on previous icon click AND on left arrow keyboard button push", async () => {
    const wrapper = mount(CardDisplay, { store, localVue });
    const fetchPrevDisplayCardSpy = jest.spyOn(
      wrapper.vm,
      "fetchPrevDisplayCard"
    );
    await flushPromises();

    await wrapper.find(".l-prev").trigger("click");
    expect(fetchPrevDisplayCardSpy).toHaveBeenCalledTimes(1);

    window.dispatchEvent(new KeyboardEvent("keyup", { key: "ArrowLeft" }));
    expect(fetchPrevDisplayCardSpy).toHaveBeenCalledTimes(2);
  });

  it("should call the |fetchNextDisplayCard| method, on next icon click AND on right arrow keyboard button push", async () => {
    const wrapper = mount(CardDisplay, { store, localVue });
    const fetchNextDisplayCardSpy = jest.spyOn(
      wrapper.vm,
      "fetchNextDisplayCard"
    );

    await wrapper.find(".l-next").trigger("click");
    expect(fetchNextDisplayCardSpy).toHaveBeenCalledTimes(1);

    window.dispatchEvent(new KeyboardEvent("keyup", { key: "ArrowRight" }));
    expect(fetchNextDisplayCardSpy).toHaveBeenCalledTimes(2);
  });
});
