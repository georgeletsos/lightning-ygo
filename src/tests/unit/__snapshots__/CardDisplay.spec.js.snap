// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDisplay.vue should render correctly 1`] = `
<div
  class="h-100"
>
  <div
    id="l-card-image"
  >
    <svg
      aria-hidden="true"
      class="l-prev animated infinite bounceLeft faster svg-inline--fa fa-caret-left fa-w-6 fa-3x"
      data-icon="caret-left"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 192 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        fill="currentColor"
      />
    </svg>
     
    <svg
      aria-hidden="true"
      class="l-next animated infinite bounceRight faster svg-inline--fa fa-caret-right fa-w-6 fa-3x"
      data-icon="caret-right"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 192 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        fill="currentColor"
      />
    </svg>
     
    <transition-stub
      enteractiveclass=""
      mode="out-in"
    >
      <img
        alt="Chimeratech Overdragon"
        class="img-fluid l-image-big"
        src="https://res.cloudinary.com/georgeletsos/image/upload/v1584279250/lightning_ygo_api/card_images/big/64599569.jpg"
      />
    </transition-stub>
  </div>
   
  <div
    id="l-card-data"
  >
    <div
      id="l-card-name-level-attr"
    >
      <div
        class="l-card-info-box"
      >
        <div
          class="l-card-info-inner-box"
        >
          <div
            class="card-name"
          >
            Chimeratech Overdragon
          </div>
           
          <hr />
           
          <div
            class="d-flex"
          >
            <div
              class="l-monster-level"
            >
              <img
                alt="Level"
                class="img-fluid"
                src=""
              />
              9
            
            </div>
             
            <div
              class="l-monster-attr"
            >
              <img
                alt="Attribute"
                class="img-fluid"
                src=""
              />
              DARK
            
            </div>
          </div>
        </div>
      </div>
    </div>
     
    <div
      id="l-card-type-text-stats"
    >
      <div
        class="l-card-info-box"
      >
        <div
          class="l-card-info-inner-box"
        >
          <div
            class="l-monster-types"
          >
            
            [Machine/Fusion/Effect]
          
          </div>
           
          <div
            class="l-card-text"
          >
            
            "Cyber Dragon" + 1+ Machine monsters
Must be Fusion Summoned. If this card is Fusion Summoned: Send all other cards you control to the GY. The original ATK/DEF of this card each become equal to the number of Fusion Materials used for its Fusion Summon x 800. Each turn, this card can attack your opponent's monsters a number of times equal to the number of Fusion Materials used for its Fusion Summon.
          
          </div>
           
          <div
            class="l-monster-stats"
          >
            
            ATK/? DEF/?
          
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
